Improve Log Message Reporting

yitzchak opened the pull request:
This a PR to improve the log reporting UI. Various ideas are listed below.

- [x] Use the new Docks API
- [x] Use events to make message display real-time (not cached until end of build)
- [x] Add message count indicator
- [x] Update busy indicator and fix the horrible wobble of the icon
- [x] Refresh log indicators on `loggingLevel` setting change
- [x] Use table for message display
- [x] Mark error locations in log files
- [x] ~~Allow logging level to be set from the dock~~ Move to filtering UI
- [x] ~~Add message copy~~ Move to separate PR
- [x] ~~Add message filtering UI (#359)~~ Move to separate PR
- [x] ~~Retain messages from previous builds~~ requires changes to log message format
- [x] Serialize messages

yitzchak commented:
@thomasjo Check the attached screen capture showing usage of the Docks API, message counter indicator and updating the log in response to changes in `loggingLevel`.
![screen capture](https://cloud.githubusercontent.com/assets/1247961/26220610/4a224dd6-3be1-11e7-8659-0a695580f883.gif)

yitzchak commented:
:fire: wobble in busy indicator.

![peek 2017-05-19 08-09](https://cloud.githubusercontent.com/assets/1247961/26247172/9d4807c4-3c6a-11e7-988c-14f725178ad8.gif)

yitzchak commented:
Table based display of messages.

![screenshot from 2017-05-21 11-37-31](https://cloud.githubusercontent.com/assets/1247961/26286409/fafe6006-3e32-11e7-82b5-1811dd60a57e.png)

yitzchak commented:
Screenshot showing `loggingLevel` controls in dock.

![peek 2017-05-23 09-25](https://cloud.githubusercontent.com/assets/1247961/26356504/40035aa4-3f9a-11e7-93e0-375aa0c72864.gif)

thomasjo commented:
> Screenshot showing loggingLevel controls in dock.

I like the intent, but from a user experience perspective it's not clear what the buttons do, nor what actually happens when you click them. Maybe a tooltip is enough, but this likely needs a label or something near the buttons that conveys "Adjust logging level". Couple that with a descriptive tooltip that clearly states what logging level here means.

What do you think? :thought_balloon:

yitzchak commented:
Agree on log filter UI. Will probably need multiple iterations so I am moving into separate PR.

yitzchak commented:
@thomasjo I've removed the logging level buttons and done some code cleanup. I think this one should ready to go.

thomasjo commented on lib/views/message-count.js:
```diff
@@ -0,0 +1,44 @@
+    const loggingLevel = atom.config.get('latex.loggingLevel')
+
+    if ((loggingLevel === 'error' && this.properties.type !== 'error') ||
+      (loggingLevel === 'warning' && this.properties.type === 'info')) {
```
Maybe we can extract two booleans here with descriptive names to make it a little bit clearer what we're checking?

thomasjo commented on lib/views/status-label.js:
```diff
@@ -17,33 +18,23 @@ export default class StatusLabel {
-    ]
-
-    if (this.properties.spin) iconClassNames.push('latex-spin')
+    if (this.properties.busy) classes.push('is-busy')
```
Pushing to a constant; this might break at some point in the future.

thomasjo commented on lib/views/status-label.js:
```diff
@@ -17,33 +18,23 @@ export default class StatusLabel {
-      <div className={classNames.join(' ')} onclick={this.properties.onClick}>
-        {this.properties.icon ? <div className={iconClassNames.join(' ')} /> : ''}
-        <span>{this.properties.text}</span>
+      <div className={classes.join(' ')} onclick={() => latex.log.show()}>
```
Perhaps introduce a method along the lines of `getClassNames()` or some such. That would also circumvent the problem above, altering a constant array.

yitzchak commented:
The test coverage went down by one because `getMostSevereType` is no longer being used. I used CSS specificity instead to make the error markers in the gutter override the warning markers, etc. The advantage of this method is that if two message ranges overlap then the non-overlapping regions will still be styled correctly.

Aside from that, I added a few more tests and fixed the issues you noted. Once I get a green light from CI I'll probably pull the trigger! :tada:

yitzchak commented:
@thomasjo I've managed to recover most of the tests that are still applicable. I'll take a look at specs for `MarkerManager` tomorrow. I think that will wrap it up.

yitzchak commented:
Using Atom's loading indicator instead?

![peek 2017-06-01 09-24](https://cloud.githubusercontent.com/assets/1247961/26681821/66ce6138-46ac-11e7-8431-12da1ea0ed69.gif)
